src/pal/src/misc/environ.cpp

```diff
@@ -634,11 +661,422 @@ SetEnvironmentVariableA(
             goto done;
         }
     }
+
     bRet = TRUE;
+
 done:
-    LOGEXIT("SetEnvironmentVariableA returning BOOL %d\n", bRet);    
+    LOGEXIT("SetEnvironmentVariableA returning BOOL %d\n", bRet);
     PERF_EXIT(SetEnvironmentVariableA);
     return bRet;
 }
 
+/*++
+Function:
+  ResizeEnvironment
+
+Resizes the PAL environment buffer.
+
+Parameters
+
+    newSize 
+           [in] New size of palEnvironment
+
+Return Values
+
+    TRUE on success, FALSE otherwise
+
+--*/
+BOOL ResizeEnvironment(int newSize)
+{
+    CPalThread * pthrCurrent = InternalGetCurrentThread();
+    InternalEnterCriticalSection(pthrCurrent, &gcsEnvironment);
+
+    if (newSize < palEnvironmentCount)
+    {
+        ASSERT("ResizeEnvironment: New size < current count!\n");
+        return FALSE;
+    }
+
+    palEnvironmentCapacity = newSize;
+
+    // If palEnvironment is null, realloc acts like malloc.
+    palEnvironment = (char**)realloc(palEnvironment, palEnvironmentCapacity * sizeof(char *));
+    if (!palEnvironment)
+    {
+        return FALSE;
+    }
+
+    InternalLeaveCriticalSection(pthrCurrent, &gcsEnvironment);
+
+    return TRUE;
+}
+
+/*++
+Function:
+  EnvironUnsetenv
+
+Remove the environment variable with the given name from the PAL version
+of the environment if it exists.
+
+Parameters
+
+    name 
+           [in] Name of variable to unset.
+
+--*/
+void EnvironUnsetenv(const char *name)
+{
+    int nameLength = strlen(name);
+
+    CPalThread * pthrCurrent = InternalGetCurrentThread();
+    InternalEnterCriticalSection(pthrCurrent, &gcsEnvironment);
+
+    for (int i = 0; palEnvironment[i] != nullptr; ++i)
+    {
+        const char *equalsSignPosition = strchr(palEnvironment[i], '=');
+        if (equalsSignPosition == nullptr)
+        {
+            equalsSignPosition = palEnvironment[i] + strlen(palEnvironment[i]);
+        }
+
+        // Check whether the name of this variable has the same length as the one
+        // we're looking for before proceeding to compare them.
+        if (equalsSignPosition - palEnvironment[i] == nameLength)
+        {
+            if (memcmp(name, palEnvironment[i], nameLength) == 0)
+            {
+                // Free the string we're removing.
+                InternalFree(palEnvironment[i]);
+
+                // Move the last environment variable pointer here.
+                palEnvironment[i] = palEnvironment[palEnvironmentCount - 1];
+                palEnvironment[palEnvironmentCount - 1] = nullptr;
+
+                palEnvironmentCount--;
```

Comment:
I left this behavior the same as it was before. If I can confirm that it's not intentional and there are no cases where we'd need to unset multiple matches, I'll add a break.